File: apps/api/app/services/permissions.py
"""Role + permission helpers — single place to ask "what can this user do?".

The rest of the codebase should call these helpers rather than comparing
`user.role` against specific enum values directly. That way adding new roles
(or remapping permissions when the spec evolves) is a single-file change.

Coverage today:
  - is_dsp_role / is_vendor_role / is_platform_role  — taxonomy buckets
  - is_org_admin                                     — authorized to manage org
                                                       (billing, users, vehicles)
  - can_invite_role                                  — the invitation matrix
                                                       enforced in
                                                       routes/invitations.py
  - allowed_invite_roles                             — UI-side enumeration
  - vendor_allowed_repair_types                      — set of RepairType
                                                       values a vendor user
                                                       can see (drives the
                                                       inspection-report
                                                       defect filter etc.)
  - defect_group_allowed_for_repair_types            — pure mapping: does
                                                       a defect's group
                                                       resolve to one of
                                                       the allowed
                                                       repair_types?

The full feature-by-feature permission map (e.g. who can approve a defect,
who can assign a WO) will be added here as those features land — keep the
checks in a single module so they're easy to audit.
"""
from __future__ import annotations

from sqlalchemy.ext.asyncio import AsyncSession
from sqlmodel import select

from app.models.user import User, UserRole


# ─────────────────────────────────────────────────────
# Taxonomy buckets
# ─────────────────────────────────────────────────────
_DSP_ROLES = frozenset({
    UserRole.DSP_OWNER,
    UserRole.DSP_MANAGER,
    UserRole.DSP_INSPECTOR,
    UserRole.DSP_VIEWER,
})

_VENDOR_ROLES = frozenset({
    UserRole.VENDOR_ADMIN,
    UserRole.SERVICE_WRITER,
    UserRole.TECHNICIAN,
    UserRole.VENDOR_VIEWER,
})

_PLATFORM_ROLES = frozenset({UserRole.SITE_ADMIN})

# Roles that have org-level admin authority — full control of users +
# fleet/inspections/WOs. Used wherever the legacy code asked for
# `dsp_owner` or `vendor_admin`.
_ORG_ADMIN_ROLES = frozenset({
    UserRole.DSP_OWNER,
    UserRole.DSP_MANAGER,
    UserRole.VENDOR_ADMIN,
    UserRole.SERVICE_WRITER,
    UserRole.SITE_ADMIN,
})


def _coerce(role) -> UserRole | None:
    """Accept either a UserRole or the raw VARCHAR value."""
    if role is None:
        return None
    if isinstance(role, UserRole):
        return role
    try:
        return UserRole(role)
    except ValueError:
        return None


def is_dsp_role(role) -> bool:
    return _coerce(role) in _DSP_ROLES


def is_vendor_role(role) -> bool:
    return _coerce(role) in _VENDOR_ROLES


def is_platform_role(role) -> bool:
    return _coerce(role) in _PLATFORM_ROLES


def is_org_admin(user_or_role) -> bool:
    """True for org-level admins (owner / manager / service_writer / site_admin).

    Accepts either a `User` instance or a raw role value/enum so callers
    don't have to unwrap `user.role` themselves.
    """
    if isinstance(user_or_role, User):
        role = user_or_role.role
    else:
        role = user_or_role
    return _coerce(role) in _ORG_ADMIN_ROLES


# ─────────────────────────────────────────────────────
# Invitation matrix
# ─────────────────────────────────────────────────────
# Who can invite which roles. Site admin is special-cased — they can invite
# anyone, anywhere (existing or new org).
_INVITE_MATRIX: dict[UserRole, frozenset[UserRole]] = {
    UserRole.DSP_OWNER: frozenset({
        UserRole.DSP_OWNER,
        UserRole.DSP_MANAGER,
        UserRole.DSP_INSPECTOR,
        UserRole.DSP_VIEWER,
    }),
    UserRole.DSP_MANAGER: frozenset({
        UserRole.DSP_INSPECTOR,
        UserRole.DSP_VIEWER,
    }),
    UserRole.VENDOR_ADMIN: frozenset({
        UserRole.VENDOR_ADMIN,
        UserRole.SERVICE_WRITER,
        UserRole.TECHNICIAN,
        UserRole.VENDOR_VIEWER,
    }),
    UserRole.SERVICE_WRITER: frozenset({
        UserRole.TECHNICIAN,
        UserRole.VENDOR_VIEWER,
    }),
    UserRole.SITE_ADMIN: frozenset(UserRole),  # everyone
    # Inspectors / technicians / viewers — empty set (cannot invite anyone)
    UserRole.DSP_INSPECTOR: frozenset(),
    UserRole.DSP_VIEWER: frozenset(),
    UserRole.TECHNICIAN: frozenset(),
    UserRole.VENDOR_VIEWER: frozenset(),
}


def can_invite_role(inviter_role, target_role) -> bool:
    """True iff `inviter_role` is allowed to send an invitation for `target_role`.

    Org-scope is enforced separately by the route (the inviter can't add
    users to other orgs unless they're a site_admin) — this function only
    answers the role-membership question.
    """
    inv = _coerce(inviter_role)
    tgt = _coerce(target_role)
    if inv is None or tgt is None:
        return False
    return tgt in _INVITE_MATRIX.get(inv, frozenset())


def allowed_invite_roles(inviter_role) -> list[UserRole]:
    """List of roles `inviter_role` may target. Empty if the inviter can't invite."""
    inv = _coerce(inviter_role)
    if inv is None:
        return []
    return sorted(_INVITE_MATRIX.get(inv, frozenset()), key=lambda r: r.value)


def can_invite_anyone(inviter_role) -> bool:
    return bool(allowed_invite_roles(inviter_role))


# ─────────────────────────────────────────────────────
# Vendor scope by repair_type
# ─────────────────────────────────────────────────────
# A vendor only services certain repair buckets (Dulles Midas does
# mechanical/pm/cnmr; a body shop does body; etc.). Inspection reports +
# heatmap counts must filter defects to the vendor's catalogue — a
# mechanical-only vendor shouldn't see body defects on a van they QC'd.
# DSP + site_admin users get a `None` sentinel (no filter applied).

async def vendor_allowed_repair_types(
    session: AsyncSession, user: User
) -> set[str] | None:
    """Repair types the requesting vendor user can see.

    Returns:
        None  — for non-vendor users (DSP / site_admin → no filter).
        set[str] — repair_type values (e.g. {"mechanical","pm","cnmr"})
                   collected across all of the vendor org's workshops.
                   Empty set means "vendor org has no workshop services
                   configured" — caller should treat that as "see nothing".
    """
    if not is_vendor_role(user.role):
        return None
    if user.organization_id is None:
        return set()
    # Local import — VendorWorkshop pulls a chunk of WO V2 models; keeping
    # it lazy avoids inflating import time for callers that never touch it.
    from app.models.work_orders import VendorWorkshop

    rows = (
        await session.execute(
            select(VendorWorkshop.repair_types).where(
                VendorWorkshop.organization_id == user.organization_id
            )
        )
    ).all()
    allowed: set[str] = set()
    for (rt_list,) in rows:
        if rt_list:
            allowed.update(rt_list)
    return allowed


def defect_group_allowed_for_repair_types(
    group: str | None, allowed_repair_types: set[str]
) -> bool:
    """True iff the defect's DefectGroup maps to one of the allowed
    repair_types.

    `group` is the string in `DefectV2Response.group` ("AMR", "Body",
    "Tires", "PM", "CNMR", "Detailing", "Netradyne") — already derived
    by the defects route from `DefectRule.group`. We re-use the same
    GROUP→RepairType map the wo_bundler uses for routing so a vendor
    sees exactly the defects they could be routed.

    Unknown / missing group defaults to MECHANICAL (matches
    `wo_bundler._resolve_repair_type`'s safe-default behaviour).
    """
    # Local import — wo_bundler imports a lot; lazy keeps it cheap.
    from app.services.wo_bundler import _GROUP_TO_REPAIR_TYPE
    from app.models.work_orders import RepairType

    repair_type = _GROUP_TO_REPAIR_TYPE.get(group or "", RepairType.MECHANICAL)
    rt_value = (
        repair_type.value if hasattr(repair_type, "value") else str(repair_type)
    )
    return rt_value in allowed_repair_types

File: nova-fora-demo/src/lib/permissions.js
/**
 * Frontend-side role + permission helpers — mirror of
 * `apps/api/app/services/permissions.py`. Components should call these
 * helpers instead of comparing `user.role` against literals so adding a
 * new role is a single-file change here.
 *
 * Source of truth for the matrix lives on the backend; this file exists so
 * the UI can hide buttons/tabs the user couldn't use anyway, NOT to enforce
 * security (the API guards every mutation).
 */

// ── Taxonomy buckets ────────────────────────────────────────
const DSP_ROLES = new Set([
  'dsp_owner', 'dsp_manager', 'dsp_inspector', 'dsp_viewer',
]);

const VENDOR_ROLES = new Set([
  'vendor_admin', 'service_writer', 'technician', 'vendor_viewer',
]);

const PLATFORM_ROLES = new Set(['site_admin']);

// Org-level admins — full control of users + fleet/inspections/WOs. Use this
// in place of `user.role === 'dsp_owner' || user.role === 'vendor_admin'`.
const ORG_ADMIN_ROLES = new Set([
  'dsp_owner', 'dsp_manager',                    // DSP admin tier
  'vendor_admin', 'service_writer',              // Vendor admin tier
  'site_admin',                                  // Platform
]);

// Roles that can RUN inspections (DVIC walkaround). Inspectors are the
// canonical case; vendor techs do post-repair inspections; managers cover
// for inspectors when needed.
const CAN_INSPECT_ROLES = new Set([
  'dsp_owner', 'dsp_manager', 'dsp_inspector',
  'vendor_admin', 'technician',
  'site_admin',
]);

// Roles that can APPROVE/dismiss reported defects (the DSP-side gate
// before a defect becomes a work order). Inspector reports them; manager
// or owner approves.
const CAN_APPROVE_DEFECTS_ROLES = new Set([
  'dsp_owner', 'dsp_manager',
  'site_admin',
]);

// Roles that can manage VEHICLES (add/edit/delete + bulk upload). DSP-side.
const CAN_MANAGE_VEHICLES_ROLES = new Set([
  'dsp_owner', 'dsp_manager',
  'site_admin',
]);

// Roles that can ACT on work orders (assign, accept, decline, complete).
// Vendor-side — DSPs see WO status but don't move them through stages.
const CAN_MANAGE_WO_ROLES = new Set([
  'vendor_admin', 'service_writer', 'technician',
  'site_admin',
]);

// Read-only roles. Useful when we need to disable everything in bulk.
const VIEWER_ROLES = new Set([
  'dsp_viewer', 'vendor_viewer',
]);


// ── Coercion helper ─────────────────────────────────────────
function _role(userOrRole) {
  if (userOrRole == null) return null;
  if (typeof userOrRole === 'string') return userOrRole;
  return userOrRole.role || null;
}


// ── Public API ──────────────────────────────────────────────
export const isDspRole = (u) => DSP_ROLES.has(_role(u));
export const isVendorRole = (u) => VENDOR_ROLES.has(_role(u));
export const isPlatformRole = (u) => PLATFORM_ROLES.has(_role(u));

export const isOrgAdmin = (u) => ORG_ADMIN_ROLES.has(_role(u));
export const isViewer = (u) => VIEWER_ROLES.has(_role(u));

export const canInspect = (u) => CAN_INSPECT_ROLES.has(_role(u));
export const canApproveDefects = (u) => CAN_APPROVE_DEFECTS_ROLES.has(_role(u));
export const canManageVehicles = (u) => CAN_MANAGE_VEHICLES_ROLES.has(_role(u));
export const canManageWorkOrders = (u) => CAN_MANAGE_WO_ROLES.has(_role(u));

/**
 * Which "view mode" the fleet/dashboard pages should render in.
 *   'owner'      — DSP-owner-style: their own vans, full edit
 *   'dsp_reader' — DSP read-only (inspector/viewer)
 *   'vendor'     — vendor-side: assigned DSPs, WO management
 *   'admin'      — site admin god mode
 */
export function getViewMode(u) {
  const r = _role(u);
  if (r === 'site_admin') return 'admin';
  if (r === 'dsp_owner' || r === 'dsp_manager') return 'owner';
  if (r === 'dsp_inspector' || r === 'dsp_viewer') return 'dsp_reader';
  if (r === 'vendor_admin' || r === 'service_writer'
      || r === 'technician' || r === 'vendor_viewer') return 'vendor';
  return 'owner';  // safe default
}


// ── Invitation matrix (mirrors backend) ─────────────────────
export const INVITE_MATRIX = {
  site_admin: [
    'dsp_owner', 'dsp_manager', 'dsp_inspector', 'dsp_viewer',
    'vendor_admin', 'service_writer', 'technician', 'vendor_viewer',
    'site_admin',
  ],
  dsp_owner:      ['dsp_owner', 'dsp_manager', 'dsp_inspector', 'dsp_viewer'],
  dsp_manager:    ['dsp_inspector', 'dsp_viewer'],
  vendor_admin:   ['vendor_admin', 'service_writer', 'technician', 'vendor_viewer'],
  service_writer: ['technician', 'vendor_viewer'],
  // empty — not allowed to invite
  dsp_inspector:  [],
  dsp_viewer:     [],
  technician:     [],
  vendor_viewer:  [],
};

export function canInviteRole(inviter, target) {
  return (INVITE_MATRIX[_role(inviter)] || []).includes(target);
}

export function allowedInviteRoles(inviter) {
  return INVITE_MATRIX[_role(inviter)] || [];
}


// ── Friendly labels ─────────────────────────────────────────
export const ROLE_LABELS = {
  dsp_owner:      'DSP Owner',
  dsp_manager:    'DSP Manager',
  dsp_inspector:  'DSP Inspector',
  dsp_viewer:     'DSP Viewer',
  vendor_admin:   'Vendor Admin',
  service_writer: 'Service Writer',
  technician:     'Technician',
  vendor_viewer:  'Vendor Viewer',
  site_admin:     'Site Admin',
};

export const ROLE_DESCRIPTIONS = {
  dsp_owner:      'Full admin: billing, users, fleet, inspections, work orders.',
  dsp_manager:    'Manages fleet + WOs. Cannot manage users or billing.',
  dsp_inspector:  'Runs DVIC inspections + reports defects. Read-only on WOs.',
  dsp_viewer:     'Read-only across the DSP.',
  vendor_admin:   'Full admin: billing, users, WO acceptance, technician assignment.',
  service_writer: 'Receives WOs, assigns technicians, talks with the DSP.',
  technician:     'Picks up assigned WOs, marks progress, completes work.',
  vendor_viewer:  'Read-only across the vendor.',
  site_admin:     'Nova Fora team — full system access.',
};
